Bugsnag/Delivery/BSGEventUploader.m

```diff
@@ -82,7 +92,68 @@ - (void)uploadEvent:(BugsnagEvent *)event completionHandler:(nullable void (^)(v
     [self.uploadQueue addOperation:operation];
 }
 
++ (void)synchronouslyUploadAtomicReportsWithConfiguration:(BugsnagConfiguration *)configuration {
+
+    BugsnagNotifier *notifier = [BugsnagNotifier new];
+    NSFileManager *fm = NSFileManager.defaultManager;
+    NSError *error = nil;
+
+    NSString *container = [BSGFileLocations atomicDirectoryContainer];
+
+    // If the parent directory does not exist, there is nothing to do.
+    if (![fm fileExistsAtPath:container]) {
+        return;
+    }
+
+    NSArray<NSString *> * contents = [fm contentsOfDirectoryAtPath:container error:&error];
+    if (!contents) {
+        bsg_log_err(@"failed to get contents of atomic container: %@", error);
+        return;
+    }
+
+    // Enumerate subdirectories.
+    for (NSString *item in contents) {
+        NSString *fullItemPath = [container stringByAppendingPathComponent:item];
+        BOOL isDirectory;
+        if (![fm fileExistsAtPath:fullItemPath isDirectory:&isDirectory] || !isDirectory) {
+            continue;
+        }
+
+        // Instantiate locations for this subdirectory.
+        BSGFileLocations *locations = [[BSGFileLocations alloc] initWithSubdirectory:item];
```

Comment:
Seems a better approach would be to give `BSGFileLocations` behaviour whereby it fails to init if `item` doesn't exist, or isn't a directory, rather than trying to do a pre-emptive check of file exists here.

Reply:
Okidoke. One other thing you could consider is the URL-based API for getting the contents of a directory. As part of that, tell it to preload whether the URLs are themselves directories. Query that resource property of each URL. Should be neater still, and means fewer trips to disk in theory.